### ISSUE TITLE
Redesign when ?

### ISSUE THREAD

**Issue by fabricerosay** [jonathan-laurent/AlphaZero.jl#217]
I played a little bit with the dev branch of redesign by GSOC author. It is a bit slower than AlphaGPU I think (pure performance wise) but with a few adjustement to Gumbel it works impressively well: this is with 32 rollouts and 4 actions, 20000 env on 3080, after around 6 minutes it catches AlphaZero.jl and then gets slightly better on tests.
![pascal_pons_benchmark_error_rates](https://github.com/user-attachments/assets/50c37c3e-b619-41f1-a6bb-a5b837f4f2a1)

**Comment by jonathan-laurent** [comment]
Yes, @AndrewSpano did some really good work on this. I still haven't had time to finish the redesign but I am not losing hope! You mentioned adjustments to Gumbel: if you changed anything, would you mind submitting a PR?

**Comment by fabricerosay** [comment]
Essentially you have to use the improved gumbel policy as training target (not the sequential halving move one hot encoded) and use the policy for moving during selfplay (again not the move you get from search), and add diversity at starting position (when reseting env, it randomly choose a position in all the positions you get with 0, 1 or two ply). My code is messy and naming is poor, here are essentially the 3 function to add/chnage in BatchedMCTS
```julia
"""
    gumbel_policy(tree, mcts_config, gumbel)

Returns an array of size (num_envs,) containing the resulting actions selected
by the sequential halving procedure with gumbel for each environment. This function should
be used after `gumbel_explore()` has been run.
"""
function gumbel_policy(tree, mcts_config, current_steps,rng::AbstractRNG)
    num_actions = Val(n_actions(tree))
#    τ = mcts_config.tau
 #   deterministic_move_idx = mcts_config.collapse_tau_move
    c_scale, c_visit = mcts_config.value_scale, mcts_config.max_visit_init
    probs = DeviceArray(mcts_config.device)(rand(rng, Float32, batch_size(tree)))
    actions = zeros(Int16, mcts_config.device, batch_size(tree))
    Devices.foreach(1:batch_size(tree), mcts_config.device) do bid
        # if current_steps[bid] >= 30#deterministic_move_idx 
        #     t=0.3f0
        # else
        #     t=1
        # end
        policy=get_ipolicy(tree,c_scale,c_visit,bid,num_actions,1.0f0)
        actions[bid] = categorical_sample(policy, probs[bid])#gumbel_mcts_action(c_scale, c_visit, tree, bid, gumbel, num_actions)
    end

    return actions
end


"""
    get_root_children_visits(tree, mcts_config)

Returns an array of size (num_actions, num_envs) containing the number of visits for
each action at the root node for each environment. This function should be used after
`gumbel_explore()` or `explore()` has been run.
"""
function get_ipolicy(tree,c_scale,c_visit, bid, num_actions::Val{A},τ=1.0f0) where {A}
    logits = SVector{A}(imap(aid -> tree.valid_actions[aid, 1, bid] ? tree.logit_prior[aid, 1, bid]/τ : -Inf32, 1:A))+
    transformed_qvalues(c_scale, c_visit, tree, 1, bid, num_actions)
    return softmax(logits)
end
function get_root_ipolicy(tree, mcts_config,c_scale=0.1f0,c_visit=50)
    # compute the policy: π′ = softmax(logits + σ(completedQ))
    num_actions = Val(n_actions(tree))
    #c_scale, c_visit = mcts_config.value_scale, mcts_config.max_visit_init
    ipolicy=zeros(Float32, mcts_config.device, (n_actions(tree), batch_size(tree)))
    Devices.foreach(1:batch_size(tree), mcts_config.device) do bid
        ipolicy[:,bid] .= get_ipolicy(tree,c_scale,c_visit,bid,num_actions)
    end
    ipolicy
end
```

And in train:
```julia
if config.use_gumbel_mcts
            t = @elapsed tree, gumbel = MCTS.gumbel_explore(mcts_config, envs, mcts_rng)
            times.explore_times[step] = t

            t = @elapsed begin
                actions = MCTS.gumbel_policy(tree, mcts_config,  steps_counter)
                policy= get_root_ipolicy(tree, mcts_config)|>cpu
                policy=[SVector{7}(policy[:,k]) for k in 1:size(policy)[2]]
            end
            times.selection_times[step] = t
```
How i choose random opening is through reset env  but very ugly and not generic: but what should be done is randomly choose a number of steps and then play randomly those steps from beginning( for connect 4 i use 2 steps max)

**Comment by jonathan-laurent** [comment]
Thanks, this is very useful!

**Comment by fabricerosay** [comment]
Yes they have been automatically generated with the last version of the repo. Only the version that use one hot action as policy target needs core change: one line in ReplayBuffer to export correctly the policy.

**Comment by jonathan-laurent** [comment]
@tasuki I put the `redesign` folder on a different branch. In the short term, I'll work with @AndrewSpano and @fabricerosay to merge an updated implementation that fully works on GPU in a `full-gpu` folder on the main branch. Hopefully, this will be done within 2-3 months. In the longer run, I'll try to provide a single package that generalizes both implementations, but I cannot yet provide a timeline for this.

**Comment by jonathan-laurent** [comment]
@fabricerosay @AndrewSpano I looked at both your versions again. I think we aren't far from being able to merge (in a dedicated `full-gpu` subdirectory of the main repo). To me, the shortest path is for @AndrewSpano to first review and integrate Fabrice's improvements. Integrating those improvements is important to me since they yield significant gains, bringing performance to the level that I had set as a target when starting this project.

As a result, I would like to have a codebase that checks the following properties:

- Fabrice's improvements must be integrated.
- All experiments replicate using latest Flux and Julia 1.11 (1.12 still does not work great with Flux).
- The Blue style guide should be followed everywhere (and especially in the new additions from Fabrice). Proper docstrings should be added when appropriate. Checks must be made that all functions belong to the right module.
- Code that is currently dead or unused should be removed or at least flagged. For all remaining instances of commented out code, an explanation should be provided for what the commented out code does (when not obvious).
- Clear instructions should be provided for replicating all experiments. In particular, in each case, precise shell commands should be provided, but **also** a description of the expected output (directory structure, models, screenshot of typical logs and TF dashboard...).
- The whole codebase should be checked for grammar and typos.

Once this is done, I commit to do a full review myself within 3 weeks at most.

**Comment by jonathan-laurent** [comment]
@fabricerosay What is [`interactive_play.jl`](https://github.com/fabricerosay/AlphaZero.jl/blob/dev/redesign/examples/interactive_play.jl) doing? It seems pretty complex. Can you provide screenshots of the corresponding visualisation and maybe document it a bit better?

**Comment by fabricerosay** [comment]
I don't think i added interactive_play myself (some parts of the code are commented which is a thing i don't ever do:), i probably modified with some specific addition for nnue on codingame (the ugly function is used to encode an int8 quantized networks in an utf16 string in a compressed format, as cg bots are limited to 100k characters) so we can get rid of it. The rest is very usefull it allows to manually play against the neural network or alphazero, which for games other than connect4 for which we have solid tests to attest the training work, permits to have a first idea of the strength of the neural network trained (i use it a lot).
Also in my last version i removed Device.foreach in favor of using AcceleratedKernels. It works perfectly and avoid some allocations, it seems neutral in terms of speed. Also i added a new mechanism for reseting games: i have a startenvs similar to envs and i randomly save some position and when resetting i reset game to starting position with 50% probability and to the corresponding startposition else. This gives a huge boost to training, particularly if you inspect tensorboard blogs you'll notice that the neural net is for example very bad on end_easy positions for a very long time(even though with az it is not so sensible) (they reach around 96% accuracy on test positions without search and  when using search you get to 98.2/98.3 quite close to Oracle but in less that 40 min with one gpu)

**Comment by jonathan-laurent** [comment]
> Also in my last version i removed Device.foreach in favor of using AcceleratedKernels.

This isn't included in the repo you shared right?

> Also i added a new mechanism for reseting games

Is this included?

**Comment by fabricerosay** [comment]
No i can push it now i wanted to test first if it works.

**Comment by jonathan-laurent** [comment]
If possible, try to have separate commits and make sure the commits are small and clean. This way, @AndrewSpano and I can decide whether to integrate each change.

In the case of AcceleratedIndex, I would be particularly interested in merging if it fully removes the need for the Devices module. I do not want to have the same codebase both introduce a new dependency and still rely on the old trick.

**Comment by fabricerosay** [comment]
I'll try, but be mercyfull for the noob I am.

**Comment by fabricerosay** [comment]
I pushed a new branch for Hex (with my old bitboard implementation). It seems to work after around one hour of training on 9x9 it was able to beat mohex2.0 showing clear progress.
But things are fishy using 16 rollouts seems not better than 4, also using a big 1024x8 nets it was not working. To make it work i had to start with a low lr of 10^-4 and making it grow progressively. It seems to me there could be something not right but not sure (sometimes value loss stays for ever at 2, it could be severe overfiting: using the whole data for training is too much, doing 2 epochs make this even worse and using bigger nets is another hint, maybe smaller net and lessrollouts help combat overfiting with data being poorer)

One last comment: instead of saving states (which waste a lot of memory you could save the env (around 2 or 4 uint64) and when needded write vectorized sate on the fly. For example for training preallocate one batch and make a vectorise_state kernel taking the batch of env and writing the batch of states. This also would permit to mask not legal moves without storing anything else.

**Comment by fabricerosay** [comment]
I think I know why hex does not work: vectorized state is wrong because you can not just juxtapose vectorized state of bitboards in right order you have to apply some symmetry because board is different with player point of view. This slightly complicate things(policy has also to be modified)

**Comment by fabricerosay** [comment]
I confirm there were a problem with state encoding now it works. Still doing some experiments, before pushing on hex branch. Also tried to train a 128x5 resnet with 5000 env, performance is decent on connect 4 but it is slower that densenet. There is still a problem when going with big nets 1024x8 0.001 learning rate is too high at begining , but going for 0.00001 and gradually setting lr higher works (around 2 hours to beat mohex, with 600 rollouts, and also starting with a small 256x4 the training worked so now it is probably faster than boardlaw). Using SGD with momentum does not seems to have this problem (it seems to learn with a 0.01 lr).
As a side note alphaGPU suffer from the same flaw for state encoding, so it is probably much better than what is advertised for hex (9 hours or so on 9x9)

**Comment by fabricerosay** [comment]
I made other modifications to hex and can confirm training work correctly. I trained a net for hex 11x11. Using fully convolutionnal so that the same net net can play on any size. I start on 5x5 and after 1000 steps i make it train on 7x7 then after around 2500 steps,9x9 and then 11x11 for 10000 steps( i use 4 rollouts till 7x7 then 8, then 16). In around 12 hours(didn't mesure properly) it ranked third on codingame (slightly better in score than my previous net), using policy only.

**Comment by fabricerosay** [comment]
Also to monitor training better it would be needed to make elo mesurement between subsequent versions of the net. Not all game have Pons Benchmark:)

**Comment by fabricerosay** [comment]
After assessing training worked i trained a better architecture(fully conv resnet à la Katago, without global pooling) with same strategy only i did 10k steps with 16 rollouts and 5400 steps with 32 rollouts when hiting 11x11 size, took around 20h. Added a very crude alphazero for the search on the codingame bot (only cpu and no multithreading it reaches around 100 rollouts) and hit first place with quite a margin. The 4th bot is from the guy who coined the term MCTS and famous for his go bots. Nice feat for Redesign®.
<img width="1148" height="286" alt="Image" src="https://github.com/user-attachments/assets/2283aa03-72cf-40c4-8228-92765098edbd" />

**Comment by jonathan-laurent** [comment]
Where is this dashboard from?

**Comment by fabricerosay** [comment]
It is from https://cgstats.magusgeek.com/app/multi-multi-puzzle-3lvf4lah1n/MrSubZero very cool site to monitor progress of a bot on codingame (without it  you would need to handle every game by hand to get such statistics)
